native_client/deepspeech.cc

```diff
@@ -428,8 +494,87 @@ ModelState::decode(vector<float>& logits)
   vector<Output> out = ctc_beam_search_decoder(
     inputs.data(), n_frames, num_classes, *alphabet, beam_width,
     cutoff_prob, cutoff_top_n, scorer);
+	
+  return out;
+}
 
-  return strdup(alphabet->LabelsToString(out[0].tokens).c_str());
+vector<META_DICT>
+ModelState::metadata_from_output(Output out,int logit_count)
+{
+  vector<META_DICT> word_list;
+
+  const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+  const int n_frames = logit_count / (BATCH_SIZE * num_classes);
+
+  std::string word = "";
+  int word_start_timestep = 0;
+
+  // Loop through each character
+  for (int t=0; t<out.tokens.size(); t++) {
+    const char * character = alphabet->StringFromLabel(out.tokens[t]).c_str();
+
+	if (strcmp(character," ") != 0) word.append(character);
+				
+	// Figure out word boundaries by looking for spaces
+	if (strcmp(character," ") == 0 || t == out.tokens.size()-1) {
+	  float time_position = (duration_secs/n_frames)*word_start_timestep;
+	  float word_duration = (duration_secs/n_frames)*(out.timesteps[t] - word_start_timestep - 3); // 3ms offset
+	  if (word_duration < 0) word_duration = 0;
+
+	  META_DICT word_info;
+	  word_info["word"] = word;
+	  word_info["time"] = std::to_string(time_position);
+	  word_info["duration"] = std::to_string(word_duration);
+	  
+	  word_list.push_back(word_info);
+												
+	  // Reset
+	  word = "";		
+	  word_start_timestep = 0;
+			
+	} else {
+	  // Timesteps are from the position of the highest letter probability so we 
+	  // offset them back by 3ms to account for this
+	  if (word.length() == 1) word_start_timestep = out.timesteps[t] - 3;
+	  if (word_start_timestep < 0) word_start_timestep = 0;
+	}
+  }
+	
+  return word_list;	
+}
+
+char* 
+ModelState::json_output_from_metadata(vector<META_DICT> word_list)
+{
+  std::ostringstream out_string;
+
+  out_string << "{" << std::endl << "\t\"file\": {\"duration\":\"" << duration_secs << "\"},";
+  out_string << std::endl << "\t\"words\": [";
+
+  // Loop through each word object
+  for (int i=0; i<word_list.size(); i++) {
+    META_DICT word = word_list[i];
+
+    out_string << std::endl << "\t\t{";
+		
+    META_DICT::iterator it = word.begin();
+
+    // Output each key and value
+    while(it != word.end()) {
+	  out_string << "\"" << it->first << "\":\"" << it->second << "\"";
+      it++;
+
+      if (it != word.end()) out_string << ", ";
+    }
+		
+    out_string << "}";
+	  
+    if (i < word_list.size()-1) out_string << ",";
+  }
+
+  out_string << std::endl << "\t]" << std::endl << "}" << std::endl;
+
+  return strdup(out_string.str().c_str());
```

Comment:
JSON is really not the best output format for this API. What should I do if I'm a C++ consumer of the API and want to use the timing information? I'd have to parse the JSON back from a string again. This should really be a properly typed API, modeled after the existing `Output` structure in the decoder for example.

Reply:
:+1: let's keep this kind of complexity out for now

Reply:
I started implementing it for my use-case and found that actually JSON is pretty annoying as an output format due to the requirement to process the audio in short sections. I have to parse each output individually and then combine them all together, so it's not really a time-saver. Something simpler like CSV would be better and the end-user can put it in a JSON dictionary later if they want.